File: src/main/java/work/lclpnet/ap2/game/aim_master/PositionGenerator.java
package work.lclpnet.ap2.game.aim_master;

import net.minecraft.block.Blocks;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.BlockPos;
import work.lclpnet.ap2.impl.util.BlockBox;

import java.util.ArrayList;
import java.util.Random;

import static java.lang.Math.*;

public class PositionGenerator {

    private final int radius;
    private final int offset;
    private final int fov;
    private final int targetNumber;
    private final BlockPos center;
    private final ServerWorld world;

    public PositionGenerator(int radius, int offset, BlockPos center, ServerWorld world, int fov, int targetNumber) {
        this.radius = radius;
        this.offset = offset;
        this.center = center;
        this.world = world;
        this.fov = fov;
        this.targetNumber = targetNumber;

        ArrayList<BlockPos> blockPositions = pickPositions();
        for (BlockPos blockPos : blockPositions) world.setBlockState(blockPos, Blocks.TARGET.getDefaultState());
    }

    private BlockBox generateBlockBox() {
        BlockPos pos1 = new BlockPos(center.getX() + radius, center.getY() + radius, (center.getZ()+offset) + radius);
        BlockPos pos2 = new BlockPos(center.getX() - radius, center.getY() - radius, (center.getZ()+offset) - radius);
        return new BlockBox(pos1, pos2);
    }

    private ArrayList<BlockPos> generateCone() {
        ArrayList<BlockPos> validBlockPositions = new ArrayList<>();

        for (BlockPos pos : generateBlockBox()) {

            int x = pos.getX() - center.getX();
            int y = pos.getY() - center.getY();
            int z = pos.getZ() - (center.getZ()+offset);

            int squaredDistance = x * x + y * y + z * z;
            int radiusSquared = radius * radius;

            int e = 13;

            if ((abs(squaredDistance - radiusSquared) <= e)) {
                double a = radius;
                double b = radius * 0.42;

                double distanceToEllipse = (x * x) / (a * a) + (y * y) / (b * b);

                if (distanceToEllipse <= 1) {
                    double angle = getAngle(x, y, z);

                    if (angle <= fov) {
                        world.setBlockState(pos, Blocks.GLASS.getDefaultState());
                        validBlockPositions.add(pos.toImmutable());
                    }
                }
            }
        }
        return validBlockPositions;
    }

    private ArrayList<BlockPos> pickPositions() {
        ArrayList<BlockPos> cone = generateCone();
        ArrayList<BlockPos> BlockPositions = new ArrayList<>();
        final Random random = new Random();

        while (BlockPositions.size() < targetNumber && !cone.isEmpty()) {

            int randIndex = random.nextInt(cone.size());
            BlockPos selectedPos = cone.remove(randIndex);
            BlockPositions.add(selectedPos);

            for (BlockPos pos : new ArrayList<>(cone)) {

                double euclideanDistance = sqrt(pow(pos.getX() - selectedPos.getX(), 2) + pow(pos.getY() - selectedPos.getY(), 2) + pow(pos.getZ() - selectedPos.getZ(), 2));
                if (euclideanDistance <= 1.8) {cone.remove(pos);}
            }
        }
        return BlockPositions;
    }

    private static double getAngle(int x, int y, int z) {
        double[] vec = {x, y, z};
        double[] viewDir = {0, 0.5, 1};

        return toDegrees(acos(calculateDotProduct(vec, viewDir) / (vectorLen(vec) * vectorLen(viewDir))));
    }

    private static double calculateDotProduct(double[] vec1, double[] vec2) {
        double dotProduct = 0.0;
        for (int i = 0; i < vec1.length; i++) {
            dotProduct += vec1[i] * vec2[i];
        }
        return dotProduct;
    }

    private static double vectorLen(double[] vec) {
        return sqrt(vec[0]*vec[0] + vec[1]*vec[1] + vec[2]*vec[2]);
    }
}

File: src/minigames/aim_master/src/main/kotlin/work/lclpnet/ap2/game/aim_master/PositionGenerator.kt
package work.lclpnet.ap2.game.aim_master

import net.minecraft.core.BlockPos
import work.lclpnet.gaco.ds.BlockBox
import java.util.Random
import kotlin.math.*

class PositionGenerator(
    private val radius: Int,
    private val offset: Int,
    private val upwardTilt: Double,
    private val ellipseFactor: Double,
    private val center: BlockPos,
    private val fov: Int,
    private val targetNumber: Int,
    private val targetMinDistance: Int
) {
    private fun generateBlockBox(): BlockBox {
        val pos1 = BlockPos(center.x + radius, center.y + radius, (center.z + offset) + radius)
        val pos2 = BlockPos(center.x - radius, center.y - radius, (center.z + offset) - radius)
        return BlockBox(pos1, pos2)
    }

    private fun generateCone(): ArrayList<BlockPos> {
        val validBlockPositions = ArrayList<BlockPos>()

        for (pos in generateBlockBox()) {
            val x = pos.x - center.x
            val y = pos.y - center.y
            val z = pos.z - (center.z + offset)

            val squaredDistance = x * x + y * y + z * z
            val radiusSquared = radius * radius

            val e = radius * (radius * 0.004 + 0.84)

            if (abs(squaredDistance - radiusSquared) <= e) {
                val a = radius.toDouble()
                val b = radius * ellipseFactor

                val distanceToEllipse = (x * x) / (a * a) + (y * y) / (b * b)

                if (distanceToEllipse <= 1) {
                    val angle = getAngle(x, y, z)

                    if (angle <= fov) {
                        validBlockPositions.add(pos.immutable())
                    }
                }
            }
        }

        return validBlockPositions
    }

    fun pickPositions(): ArrayList<BlockPos> {
        val cone = generateCone()
        val blockPositions = ArrayList<BlockPos>()
        val random = Random()

        while (blockPositions.size < targetNumber && cone.isNotEmpty()) {
            val randIndex = random.nextInt(cone.size)
            val selectedPos = cone.removeAt(randIndex)
            blockPositions.add(selectedPos)

            cone.removeAll { pos ->
                val dx = (pos.x - selectedPos.x).toDouble()
                val dy = (pos.y - selectedPos.y).toDouble()
                val dz = (pos.z - selectedPos.z).toDouble()
                sqrt(dx * dx + dy * dy + dz * dz) <= targetMinDistance
            }
        }

        return blockPositions
    }

    private fun getAngle(x: Int, y: Int, z: Int): Double {
        val vec = doubleArrayOf(x.toDouble(), y.toDouble(), z.toDouble())
        val viewDir = doubleArrayOf(0.0, upwardTilt, 1.0)
        return Math.toDegrees(
            acos(dotProduct(vec, viewDir) / (vecLen(vec) * vecLen(viewDir)))
        )
    }
}

private fun dotProduct(vec1: DoubleArray, vec2: DoubleArray): Double {
    var result = 0.0
    for (i in vec1.indices) result += vec1[i] * vec2[i]
    return result
}

private fun vecLen(vec: DoubleArray): Double = sqrt(vec[0] * vec[0] + vec[1] * vec[1] + vec[2] * vec[2])
